File: apps/faithform-android/app/src/main/kotlin/io/faithform/app/ui/schedule/ScheduleModel.kt
package io.faithform.app.ui.schedule

import io.faithform.app.contract.FeedItem
import io.faithform.app.contract.MobileErrorCode
import io.faithform.app.contract.SchedulePage
import io.faithform.app.network.ApiClient
import io.faithform.app.network.ApiException
import io.faithform.app.network.MobileSuccess
import io.faithform.app.network.ProjectionCache
import io.faithform.app.storage.CachePartition
import io.faithform.app.storage.Freshness
import java.time.Instant
import java.time.YearMonth
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import kotlin.coroutines.cancellation.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.builtins.ListSerializer

/** Mirrors iOS `SchedulePhase`. */
sealed interface SchedulePhase {
    data object Loading : SchedulePhase
    data class Loaded(val items: List<FeedItem>, val isStale: Boolean) : SchedulePhase
    data object Empty : SchedulePhase
    data object OfflineNoCache : SchedulePhase
    data object Blocked : SchedulePhase
    data class Failed(val message: String) : SchedulePhase
}

/**
 * The Home Schedule calendar for one church.
 *
 * Loads every published event overlapping the visible month window, cached and
 * revalidated the same way as [io.faithform.app.ui.feed.FeedModel].
 */
class ScheduleModel(
    private val api: ApiClient,
    private val cache: ProjectionCache,
    private val churchSlug: String,
    private val partition: CachePartition,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    private val _phase = MutableStateFlow<SchedulePhase>(SchedulePhase.Loading)
    val phase: StateFlow<SchedulePhase> = _phase.asStateFlow()

    private val _displayedMonth = MutableStateFlow(YearMonth.now())
    val displayedMonth: StateFlow<YearMonth> = _displayedMonth.asStateFlow()

    private val _churchTimezone = MutableStateFlow(DEFAULT_TIMEZONE)
    val churchTimezone: StateFlow<String> = _churchTimezone.asStateFlow()

    private var etag: String? = null
    private val itemsSerializer = ListSerializer(FeedItem.serializer())

    suspend fun load() {
        loadMonth(_displayedMonth.value)
    }

    suspend fun refresh() {
        loadMonth(_displayedMonth.value)
    }

    suspend fun showMonth(month: YearMonth) {
        _displayedMonth.value = month
        loadMonth(month)
    }

    suspend fun showPreviousMonth() {
        showMonth(_displayedMonth.value.minusMonths(1))
    }

    suspend fun showNextMonth() {
        showMonth(_displayedMonth.value.plusMonths(1))
    }

    private suspend fun loadMonth(month: YearMonth) {
        val zone = runCatching { ZoneId.of(_churchTimezone.value) }.getOrElse { ZoneId.of(DEFAULT_TIMEZONE) }
        val window = ScheduleCalendar.monthWindow(month, zone)
        val cacheName = "schedule|${window.from}|${window.to}"

        cache.load(cacheName, partition, itemsSerializer)?.let { cached ->
            etag = cached.etag
            _phase.value = if (cached.value.isEmpty()) {
                SchedulePhase.Empty
            } else {
                SchedulePhase.Loaded(cached.value, isStale = false)
            }
        } ?: run {
            _phase.value = SchedulePhase.Loading
        }

        try {
            val response = api.send(
                path = "api/mobile/v1/schedule/$churchSlug",
                serializer = MobileSuccess.serializer(SchedulePage.serializer()),
                query = mapOf("from" to window.from, "to" to window.to),
                ifNoneMatch = etag,
            )

            if (response.notModified) {
                (_phase.value as? SchedulePhase.Loaded)?.let {
                    _phase.value = it.copy(isStale = false)
                    return
                }
                cache.load(cacheName, partition, itemsSerializer)?.let { cached ->
                    _phase.value = if (cached.value.isEmpty()) SchedulePhase.Empty else SchedulePhase.Loaded(cached.value, isStale = false)
                }
                return
            }

            val page = response.value ?: return
            etag = response.etag
            page.items.firstOrNull()?.churchTimezone?.takeIf { it.isNotBlank() }?.let {
                _churchTimezone.value = it
            }

            cache.store(cacheName, partition, itemsSerializer, page.items, response.etag)
            _phase.value = if (page.items.isEmpty()) {
                SchedulePhase.Empty
            } else {
                SchedulePhase.Loaded(page.items, isStale = false)
            }
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (error: ApiException) {
            when {
                error.code == MobileErrorCode.BLOCKED -> {
                    cache.purge(partition)
                    _phase.value = SchedulePhase.Blocked
                }
                _phase.value is SchedulePhase.Loaded -> Unit
                error.code == MobileErrorCode.UNAVAILABLE || error.retryable ->
                    _phase.value = SchedulePhase.OfflineNoCache
                else -> _phase.value = SchedulePhase.Failed(error.displayMessage)
            }
        } catch (_: Exception) {
            if (_phase.value !is SchedulePhase.Loaded) {
                _phase.value = SchedulePhase.OfflineNoCache
            }
        }
    }

    private companion object {
        const val DEFAULT_TIMEZONE = "America/New_York"
        const val TTL_MILLIS = 300_000L
    }
}

object ScheduleCalendar {
    data class Window(val from: String, val to: String)

    private val instantFormatter = DateTimeFormatter.ISO_INSTANT

    fun monthWindow(month: YearMonth, zone: ZoneId): Window {
        val start = month.atDay(1).atStartOfDay(zone).toInstant()
        val end = month.plusMonths(1).atDay(1).atStartOfDay(zone).toInstant()
        return Window(from = instantFormatter.format(start), to = instantFormatter.format(end))
    }

    fun monthTitle(month: YearMonth): String =
        month.atDay(1).format(DateTimeFormatter.ofPattern("MMMM yyyy"))

    fun eventOverlaps(item: FeedItem, day: java.time.LocalDate, zone: ZoneId): Boolean {
        val start = runCatching { Instant.parse(item.startAt).atZone(zone) }.getOrNull() ?: return false
        val dayStart = day.atStartOfDay(zone)
        val dayEnd = day.plusDays(1).atStartOfDay(zone)

        val end = item.endAt?.let { runCatching { Instant.parse(it).atZone(zone) }.getOrNull() }
            ?: if (item.allDay) {
                start.plusDays(2)
            } else {
                start.plusDays(1)
            }

        return start.isBefore(dayEnd) && end.isAfter(dayStart)
    }

    fun eventsOn(day: java.time.LocalDate, items: List<FeedItem>, zone: ZoneId): List<FeedItem> =
        items.filter { eventOverlaps(it, day, zone) }
            .sortedWith(compareByDescending<FeedItem> { it.isPinned }.thenBy { it.startAt })

    fun daysWithEvents(items: List<FeedItem>, zone: ZoneId): Set<java.time.LocalDate> {
        val days = mutableSetOf<java.time.LocalDate>()
        for (item in items) {
            val start = runCatching { Instant.parse(item.startAt).atZone(zone) }.getOrNull() ?: continue
            val end = item.endAt?.let { runCatching { Instant.parse(it).atZone(zone) }.getOrNull() }
                ?: if (item.allDay) start.plusDays(2) else start.plusDays(1)
            var cursor = start.toLocalDate()
            val last = end.toLocalDate()
            while (!cursor.isAfter(last)) {
                days.add(cursor)
                cursor = cursor.plusDays(1)
            }
        }
        return days
    }
}

File: apps/faithform-ios/Sources/FaithFormKit/Features/ScheduleModel.swift
import Foundation
import Observation

/// The Home Schedule pane's state.
public enum SchedulePhase: Equatable, Sendable {
    case loading
    case loaded(items: [FeedItem], isStale: Bool)
    case empty
    case offlineNoCache
    case blocked
    case failed(String)
}

@Observable
@MainActor
public final class ScheduleModel {
    public private(set) var phase: SchedulePhase = .loading
    public private(set) var isRefreshing = false
    public private(set) var displayedMonth: Date = Date()
    public private(set) var churchTimezone: String = "America/New_York"

    private let api: APIClient
    private let cache: PartitionedCache
    private var etag: String?
    private var partition: CachePartition?
    private var churchSlug: String = ""

    public init(api: APIClient, cache: PartitionedCache) {
        self.api = api
        self.cache = cache
    }

    public func load(
        churchSlug: String,
        churchTimezone: String,
        partition: CachePartition
    ) async {
        self.churchSlug = churchSlug
        if !churchTimezone.isEmpty {
            self.churchTimezone = churchTimezone
        }
        self.partition = partition
        await loadMonth(displayedMonth)
    }

    public func refresh(churchSlug: String) async {
        self.churchSlug = churchSlug
        await loadMonth(displayedMonth)
    }

    public func showMonth(_ month: Date) async {
        displayedMonth = month
        await loadMonth(month)
    }

    public func showPreviousMonth() async {
        guard let previous = ScheduleCalendar.shiftMonth(displayedMonth, by: -1, timezone: churchTimezone) else { return }
        await showMonth(previous)
    }

    public func showNextMonth() async {
        guard let next = ScheduleCalendar.shiftMonth(displayedMonth, by: 1, timezone: churchTimezone) else { return }
        await showMonth(next)
    }

    private func loadMonth(_ month: Date) async {
        guard let partition else { return }
        let window = ScheduleCalendar.monthWindow(month, timezone: churchTimezone)
        let cacheName = "schedule|\(window.from)|\(window.to)"

        if let cached = await cache.load([FeedItem].self, name: cacheName, partition: partition) {
            phase = cached.value.isEmpty
                ? .empty
                : .loaded(items: cached.value, isStale: false)
            etag = cached.etag
        } else {
            phase = .loading
        }

        isRefreshing = true
        defer { isRefreshing = false }

        do {
            let response = try await api.send(
                "api/mobile/v1/schedule/\(churchSlug)",
                query: ["from": window.from, "to": window.to],
                ifNoneMatch: etag,
                as: SchedulePage.self
            )

            if response.notModified {
                if case let .loaded(items, _) = phase {
                    phase = .loaded(items: items, isStale: false)
                } else if let cached = await cache.load([FeedItem].self, name: cacheName, partition: partition) {
                    phase = cached.value.isEmpty ? .empty : .loaded(items: cached.value, isStale: false)
                }
                return
            }

            guard let page = response.value else { return }
            etag = response.etag
            if let timezone = page.items.first?.churchTimezone, !timezone.isEmpty {
                churchTimezone = timezone
            }

            try? await cache.store(
                CacheEntry(value: page.items, etag: response.etag, storedAt: Date()),
                name: cacheName,
                partition: partition
            )

            phase = page.items.isEmpty ? .empty : .loaded(items: page.items, isStale: false)
        } catch let error as APIError {
            if error.isCancellation { return }
            switch error.code {
            case .blocked:
                await cache.purge(partition: partition)
                phase = .blocked
            case .unavailable:
                if case let .loaded(items, _) = phase {
                    phase = .loaded(items: items, isStale: true)
                    return
                }
                phase = .offlineNoCache
            default:
                if case .loaded = phase { return }
                phase = .failed(error.displayMessage)
            }
        } catch {
            if error.isCancellation { return }
            if case let .loaded(items, _) = phase {
                phase = .loaded(items: items, isStale: true)
                return
            }
            phase = .offlineNoCache
        }
    }
}

/// Month boundaries and day matching in the church's timezone.
public enum ScheduleCalendar {
    public struct Window: Sendable {
        public let from: String
        public let to: String
    }

    public static func monthWindow(_ month: Date, timezone: String) -> Window {
        let zone = TimeZone(identifier: timezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone

        let components = calendar.dateComponents([.year, .month], from: month)
        let start = calendar.date(from: components) ?? month
        let end = calendar.date(byAdding: .month, value: 1, to: start) ?? start

        return Window(
            from: FaithFormInstant.format(start),
            to: FaithFormInstant.format(end)
        )
    }

    public static func shiftMonth(_ month: Date, by delta: Int, timezone: String) -> Date? {
        let zone = TimeZone(identifier: timezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone
        return calendar.date(byAdding: .month, value: delta, to: month)
    }

    public static func monthTitle(_ month: Date, timezone: String) -> String {
        let zone = TimeZone(identifier: timezone) ?? .current
        let formatter = DateFormatter()
        formatter.timeZone = zone
        formatter.dateFormat = "MMMM yyyy"
        return formatter.string(from: month)
    }

    public static func daysInMonth(_ month: Date, timezone: String) -> [Date] {
        let zone = TimeZone(identifier: timezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone

        let components = calendar.dateComponents([.year, .month], from: month)
        guard let start = calendar.date(from: components),
              let range = calendar.range(of: .day, in: .month, for: start) else { return [] }

        return range.compactMap { day in
            calendar.date(byAdding: .day, value: day - 1, to: start)
        }
    }

    public static func weekdayHeaders(timezone: String) -> [String] {
        let zone = TimeZone(identifier: timezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone
        calendar.locale = Locale.current

        let formatter = DateFormatter()
        formatter.timeZone = zone
        formatter.locale = calendar.locale

        return formatter.shortWeekdaySymbols
    }

    public static func leadingBlankDays(_ month: Date, timezone: String) -> Int {
        let zone = TimeZone(identifier: timezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone
        calendar.firstWeekday = calendar.firstWeekday

        let components = calendar.dateComponents([.year, .month], from: month)
        guard let start = calendar.date(from: components) else { return 0 }
        let weekday = calendar.component(.weekday, from: start)
        return (weekday - calendar.firstWeekday + 7) % 7
    }

    public static func isSameDay(_ lhs: Date, _ rhs: Date, timezone: String) -> Bool {
        let zone = TimeZone(identifier: timezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone
        return calendar.isDate(lhs, inSameDayAs: rhs)
    }

    public static func dayNumber(_ date: Date, timezone: String) -> String {
        let zone = TimeZone(identifier: timezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone
        return String(calendar.component(.day, from: date))
    }

    /// Whether an event should appear on a calendar day in the church's zone.
    public static func event(_ item: FeedItem, overlaps day: Date) -> Bool {
        let zone = TimeZone(identifier: item.churchTimezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone

        guard let start = FaithFormInstant.parse(item.startAt) else { return false }
        let dayStart = calendar.startOfDay(for: day)
        guard let dayEnd = calendar.date(byAdding: .day, value: 1, to: dayStart) else { return false }

        let end: Date
        if let endRaw = item.endAt, let parsed = FaithFormInstant.parse(endRaw) {
            end = parsed
        } else if item.allDay {
            end = calendar.date(byAdding: .day, value: 2, to: start) ?? start
        } else {
            end = calendar.date(byAdding: .day, value: 1, to: start) ?? start
        }

        return start < dayEnd && end > dayStart
    }

    public static func events(on day: Date, in items: [FeedItem]) -> [FeedItem] {
        items.filter { event($0, overlaps: day) }
            .sorted { lhs, rhs in
                if lhs.isPinned != rhs.isPinned { return lhs.isPinned && !rhs.isPinned }
                return lhs.startAt < rhs.startAt
            }
    }

    public static func daysWithEvents(_ items: [FeedItem], timezone: String) -> Set<Date> {
        let zone = TimeZone(identifier: timezone) ?? .current
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone

        var days = Set<Date>()
        for item in items {
            guard let start = FaithFormInstant.parse(item.startAt) else { continue }
            let end: Date
            if let endRaw = item.endAt, let parsed = FaithFormInstant.parse(endRaw) {
                end = parsed
            } else if item.allDay {
                end = calendar.date(byAdding: .day, value: 2, to: start) ?? start
            } else {
                end = calendar.date(byAdding: .day, value: 1, to: start) ?? start
            }

            var cursor = calendar.startOfDay(for: start)
            let last = calendar.startOfDay(for: end)
            while cursor <= last {
                days.insert(cursor)
                guard let next = calendar.date(byAdding: .day, value: 1, to: cursor) else { break }
                cursor = next
            }
        }
        return days
    }
}
